# Error vs Exception 

오류의 종류인 에러 (`Error`)와 예외 (`Exception`) 는 비슷한듯 다른 개념이다. <br>
이번 챕터에서는 프로그래밍에서 자주 마주치는 문제인 에러와 예외의 차이와 예외처리에 대해 설명한다. <br>

## Error 

![image](https://github.com/dlrkdus/CS_STUDY/assets/99721126/df2fd711-e519-43b1-be60-bee288ca617d)


Error는 하드웨어상의 심각한 문제나 예기치 않은 시스템상의 문제로 야기되는 상황을 말한다. <br>
문법적 오류나 Null 참조가 해당할 수 있고 메모리 부족이나 스택 오버플로 등도 해당된다. <br>
Error가 발생한 경우 시스템은 강제 종료될 수 있으며 개발자가 조치할 수 없다. <br>

- StackOverflowError: 호출의 깊이가 깊어지거나 재귀가 지속되어 stack overflow 발생 시 던져지는 오류
- OutOfMemoryError: JVM이 할당된 메모리의 부족으로 더이상 객체를 할당할 수 없을 때 던져지는 오류

## Exception

![image](https://github.com/dlrkdus/CS_STUDY/assets/99721126/f1f0b796-0260-4d1a-8254-ac54a6e7316f)


Exception은 프로그램 실행 중에 발생하는 예상 가능한 문제를 말한다. <br>
주로 개발자가 구현한 로직상에서 발생하며 예측할 수 있는 문제이기 때문에 개발자가 미리 발생할만한 예외에 대하여 처리 로직(`Exception Handling`)을 작성해야 한다. <br>

Error는 발생시 복구할 수 없는 _심각한 수준_ 의 오류지만 <br>
Exception은 개발자가 적절한 처리 로직을 작성하면 _수습하거나 방지_ 할 수 있는 수준의 오류이다. <br>

- NullPointerException: 객체 참조가 null인 상태에서 객체의 메소드를 호출하려고 할 때 발생합니다.
- IndexOutOfBoundsException : 배열과 유사한 자료구조(문자열, 배열, 자료구조)에서 범위를 벗어난 인덱스 번호 사용으로 발생합니다.
- IOException: 입력/출력 작업 실패 또는 인터럽트 시 발생합니다. 파일을 읽거나 쓰는 과정에서 주로 발생합니다.
- NumberFormatException: 문자열을 숫자로 변환하려고 할 때 해당 문자열이 적절한 형식이 아닐 경우 발생합니다.
- OutOfMemoryError: JVM이 객체를 생성하기 위해 더 이상 메모리를 할당할 수 없을 때 발생합니다. 이는 에러의 한 종류로, 예외 처리보다는 예방에 중점을 둬야 합니다.
- IllegalStateException : 객체의 상태가 매소드 호출에는 부적절한 경우에 발생합니다.
<hr>

Exception은 다시 두 가지로 나누어지는데, 컴파일에서 걸리는 예외와 런타임에서 걸리는 예외가 있다. <br>

- Checked Exception : 예외처리가 필수이며, 처리하지 않으면 컴파일되지 않습니다. JVM 외부와 통신(네트워크, 파일시스템 등)할 때 주로 쓰입니다.
  - RuntimeException 이외에 있는 모든 예외
  - IOException, SQLException 등
- Unchecked Exception : 컴파일 때 체크되지 않고, Runtime에 발생하는 Exception을 말합니다.
  - RuntimeException 하위의 모든 예외
  - NullPointerException, IndexOutOfBoundException 등

## Exception Handling 

![image](https://github.com/dlrkdus/CS_STUDY/assets/99721126/7539235a-8fa4-4e95-8b09-462da27431cc)


예상 가능한 예외 상황들에 대해 처리하는 핸들러이다. <br>
Exception과 Error가 상속 받는 Throwable 클래스는 예외처리를 할 수 있는 최상위 클래스이다. <br>
자바에서는 3가지 방법으로 예외를 처리할 수 있다. <br>
Java에서 모든 예외가 발생하면 ~Exception 객체를 생성한다. <br>

### try-catch `예외복구`

try catch문으로 직접 예외를 발생시키고 처리한다. 

```
final int MAX_RETRY = 100;
public Object someMethod() {
    int maxRetry = MAX_RETRY;
    while(maxRetry > 0) {
        try {
            ...
        } catch(SomeException e) {
            // 로그 출력. 정해진 시간만큼 대기한다.
        } finally {
            // 리소스 반납 및 정리 작업
        }
    }
    // 최대 재시도 횟수를 넘기면 직접 예외를 발생시킨다.
    throw new RetryFailedException();
}
```

### throws Exception `예외 회피`

발생한 예외의 책임을 호출한쪽에서 진다. (처리한다)

```
// 예시 1
public void add() throws SQLException {
    // ...생략
}

// 예시 2 
public void add() throws SQLException {
    try {
        // ... 생략
    } catch(SQLException e) {
        // 로그를 출력하고 다시 날린다!
        throw e;
    }
}
```

### try-catch + throws Exception `예외 전환`

try catch문에서 예외는 단순하게 표현하고 예외 클래스는 더 자세히 구현하는 방법. 

```
// 조금 더 명확한 예외로 던진다.
public void add(User user) throws DuplicateUserIdException, SQLException {
    try {
        // ...생략
    } catch(SQLException e) {
        if(e.getErrorCode() == MysqlErrorNumbers.ER_DUP_ENTRY) {
            throw DuplicateUserIdException();
        }
        else throw e;
    }
}

// 예외를 단순하게 포장한다.
public void someMethod() {
    try {
        // ...생략
    }
    catch(NamingException ne) {
        throw new EJBException(ne);
        }
    catch(SQLException se) {
        throw new EJBException(se);
        }
    catch(RemoteException re) {
        throw new EJBException(re);
        }
}
```


#### 출처
- https://github.com/GimunLee/tech-refrigerator/blob/master/Language/JAVA/Error%20%26%20Exception.md#error--exception
- https://github.com/devSquad-study/2023-CS-Study/blob/main/java/java_error_exception.md








 





